From 3e98dfa70289b1297b103f78a1c21a88f68d4659 Mon Sep 17 00:00:00 2001
From: Hiroshi Hatake <hatake@clear-code.com>
Date: Fri, 2 Nov 2018 09:34:33 +0000
Subject: [PATCH] Observe didcomposite

Because previous implementation causes EGL related operation conflicts and deadlock.
Composite and Resize are exclusive operation on R-Car Gen3 series' PVR driver.
So, we should avoid this deadlock with DidCompositeOserver.
Because this patch makes more determinisitic EGL operation than previous.
---
 widget/gtk/mozcontainer.cpp | 44 ++++++++++++++++++++++++++++++++++++--------
 widget/gtk/mozcontainer.h   |  6 ++++++
 widget/gtk/nsWindow.cpp     | 37 +++++++++++++++++++++++++++++++++++++
 widget/gtk/nsWindow.h       |  2 ++
 4 files changed, 81 insertions(+), 8 deletions(-)

diff --git a/widget/gtk/mozcontainer.cpp b/widget/gtk/mozcontainer.cpp
index 82fbfc6..8d240a3 100644
--- a/widget/gtk/mozcontainer.cpp
+++ b/widget/gtk/mozcontainer.cpp
@@ -213,6 +213,9 @@ moz_container_init (MozContainer *container)
       container->subsurface = nullptr;
       container->eglwindow = nullptr;
       container->committed = FALSE;
+      container->egl_window_width = -1;
+      container->egl_window_height = -1;
+
 
       GdkDisplay *gdk_display = gtk_widget_get_display(GTK_WIDGET(container));
       if (GDK_IS_WAYLAND_DISPLAY (gdk_display)) {
@@ -459,11 +462,6 @@ moz_container_size_allocate (GtkWidget     *widget,
         gdk_window_get_position(gtk_widget_get_window(widget), &x, &y);
         wl_subsurface_set_position(container->subsurface, x, y);
     }
-    if (container->eglwindow) {
-        wl_egl_window_resize(container->eglwindow,
-                             allocation->width, allocation->height,
-                             0, 0);
-    }
 #endif
 }
 
@@ -599,10 +597,13 @@ moz_container_get_wl_egl_window(MozContainer *container)
             return nullptr;
 
       GdkWindow *window = gtk_widget_get_window(GTK_WIDGET(container));
+      gint width = gdk_window_get_width(window);
+      gint height = gdk_window_get_height(window);
       container->eglwindow
-            = wl_egl_window_create(wlsurf,
-                                   gdk_window_get_width(window),
-                                   gdk_window_get_height(window));
+          = wl_egl_window_create(wlsurf, width, height);
+
+      container->egl_window_width = width;
+      container->egl_window_height = height;
     }
     return container->eglwindow;
 }
@@ -618,4 +619,31 @@ moz_container_is_committed(MozContainer *container)
 {
     return container->committed;
 }
+
+void
+moz_container_set_wl_egl_window_size(MozContainer *container,
+                                     gint width,
+                                     gint height)
+{
+    if (!container->eglwindow)
+        return;
+
+    GtkWidget *widget = GTK_WIDGET(container);
+    GtkAllocation allocation;
+    gtk_widget_get_allocation(widget, &allocation);
+
+    if (allocation.width  != container->egl_window_width ||
+        allocation.height != container->egl_window_height) {
+
+        container->egl_window_width  = allocation.width;
+        container->egl_window_height = allocation.height;
+        wl_egl_window_resize(container->eglwindow,
+                             container->egl_window_width,
+                             container->egl_window_height,
+                             0, 0);
+
+        gdk_window_invalidate_rect(gtk_widget_get_window(widget),
+                                   nullptr, FALSE);
+    }
+}
 #endif
diff --git a/widget/gtk/mozcontainer.h b/widget/gtk/mozcontainer.h
index a78f089..d12fca3 100644
--- a/widget/gtk/mozcontainer.h
+++ b/widget/gtk/mozcontainer.h
@@ -74,6 +74,9 @@ struct _MozContainer
     struct wl_subsurface    *subsurface;
     struct wl_egl_window    *eglwindow;
     gboolean committed;
+    gint egl_window_width;
+    gint egl_window_height;
+
 #endif
 };
 
@@ -100,6 +103,9 @@ struct wl_surface* moz_container_get_wl_surface(MozContainer *container);
 struct wl_egl_window* moz_container_get_wl_egl_window(MozContainer *container);
 gboolean moz_container_has_wl_egl_window(MozContainer *container);
 gboolean moz_container_is_committed(MozContainer *container);
+void     moz_container_set_wl_egl_window_size(MozContainer *container,
+                                              gint width,
+                                              gint height);
 #endif
 
 #endif /* __MOZ_CONTAINER_H__ */
diff --git a/widget/gtk/nsWindow.cpp b/widget/gtk/nsWindow.cpp
index a6a0b8c..7d11076 100644
--- a/widget/gtk/nsWindow.cpp
+++ b/widget/gtk/nsWindow.cpp
@@ -345,6 +345,32 @@ private:
     TimeStamp  mAsyncUpdateStart;
 };
 
+class nsWindowDidCompositeObserver final : public DidCompositeObserver
+{
+public:
+  nsWindowDidCompositeObserver(MozContainer *aContainer, LayerManager* aLayerManager)
+    : mContainer(aContainer),
+      mLayerManager(aLayerManager)
+  {
+  }
+
+  ~nsWindowDidCompositeObserver() {
+  }
+
+  void DidComposite() override {
+    GtkAllocation allocation;
+    GtkWidget* widget = GTK_WIDGET(mContainer);
+    gtk_widget_get_allocation (widget, &allocation);
+    moz_container_set_wl_egl_window_size(mContainer,
+                                         allocation.width,
+                                         allocation.height);
+  }
+
+private:
+  MozContainer* mContainer;
+  RefPtr<LayerManager> mLayerManager;
+};
+
 } // namespace mozilla
 
 static NS_DEFINE_IID(kCDragServiceCID,  NS_DRAGSERVICE_CID);
@@ -488,6 +514,9 @@ nsWindow::nsWindow()
     mPendingConfigures = 0;
     mCSDSupportLevel = CSD_SUPPORT_NONE;
     mDrawInTitlebar = false;
+#ifdef MOZ_WAYLAND
+    mDidCompositeObserver = nullptr;
+#endif
 }
 
 nsWindow::~nsWindow()
@@ -2089,6 +2118,14 @@ nsWindow::OnExposeEvent(cairo_t *cr)
     if (GetLayerManager()->AsKnowsCompositor() && mCompositorSession) {
         // We need to paint to the screen even if nothing changed, since if we
         // don't have a compositing window manager, our pixels could be stale.
+#ifdef MOZ_WAYLAND
+        if (mContainer) {
+            if (!mDidCompositeObserver) {
+                mDidCompositeObserver = MakeUnique<nsWindowDidCompositeObserver>(mContainer, GetLayerManager());
+                GetLayerManager()->AddDidCompositeObserver(mDidCompositeObserver.get());
+            }
+        }
+#endif
         GetLayerManager()->SetNeedsComposite(true);
         GetLayerManager()->SendInvalidRegion(region.ToUnknownRegion());
     }
diff --git a/widget/gtk/nsWindow.h b/widget/gtk/nsWindow.h
index 5a96079..7ef7804 100644
--- a/widget/gtk/nsWindow.h
+++ b/widget/gtk/nsWindow.h
@@ -86,6 +86,7 @@ class gfxPattern;
 namespace mozilla {
 class TimeStamp;
 class CurrentX11TimeGetter;
+class nsWindowDidCompositeObserver;
 }
 
 class nsWindow final : public nsBaseWidget
@@ -614,6 +615,7 @@ private:
     RefPtr<mozilla::widget::IMContextWrapper> mIMContext;
 
     mozilla::UniquePtr<mozilla::CurrentX11TimeGetter> mCurrentTimeGetter;
+    mozilla::UniquePtr<mozilla::nsWindowDidCompositeObserver> mDidCompositeObserver;
     static CSDSupportLevel sCSDSupportLevel;
 };
 
-- 
2.7.4

